framework/src/main/java/org/checkerframework/framework/type/QualifierHierarchy.java

```diff
@@ -138,298 +198,248 @@ public int numberOfIterationsBeforeWidening() {
-
-        return result;
-    }
+    // The fact that null is returned if the qualifiers are not in the same hierarchy is used by the
+    // set version of LUB below.
+    public abstract @Nullable AnnotationMirror greatestLowerBound(
+            AnnotationMirror a1, AnnotationMirror a2);
 
     /**
-     * Returns the greatest lower bound of two types. Each type is represented as a set of type
-     * qualifiers, as is the result.
-     *
-     * <p>Annos1 and annos2 must have the same size, and each annotation in them must be from a
-     * different type hierarchy.
+     * Returns the least upper bound of two the two sets of qualifiers. The result is the lub of the
+     * qualifier for the same hierarchy in each set.
      *
-     * @param annos1 first collection of qualifiers
-     * @param annos2 second collection of qualifiers
-     * @return pairwise greatest lower bounds of elements of the input collections (which need not
-     *     be sorted in the same order)
+     * @param qualifiers1 set of qualifiers; exactly one per hierarchy
+     * @param qualifiers2 set of qualifiers; exactly one per hierarchy
+     * @return the least upper bound of two the two sets of qualifiers
      */
     public Set<? extends AnnotationMirror> greatestLowerBounds(
-            Collection<? extends AnnotationMirror> annos1,
-            Collection<? extends AnnotationMirror> annos2) {
-        assertSameSize(annos1, annos2);
-        if (annos1.isEmpty()) {
+            Collection<? extends AnnotationMirror> qualifiers1,
+            Collection<? extends AnnotationMirror> qualifiers2) {
+        assertSameSize(qualifiers1, qualifiers2);
+        if (qualifiers1.isEmpty()) {
             throw new BugInCF(
                     "QualifierHierarchy.greatestLowerBounds: tried to determine GLB with empty sets");
         }
 
         Set<AnnotationMirror> result = AnnotationUtils.createAnnotationSet();
-        for (AnnotationMirror a1 : annos1) {
-            for (AnnotationMirror a2 : annos2) {
+        for (AnnotationMirror a1 : qualifiers1) {
+            for (AnnotationMirror a2 : qualifiers2) {
                 AnnotationMirror glb = greatestLowerBound(a1, a2);
                 if (glb != null) {
                     result.add(glb);
                 }
             }
         }
 
-        assertSameSize(annos1, annos2, result);
-
+        assertSameSize(qualifiers1, qualifiers2, result);
         return result;
     }
 
     /**
-     * Tests whether {@code subAnno} is a sub-qualifier of, or equal to, {@code superAnno},
-     * according to the type qualifier hierarchy. This checks only the qualifiers, not the Java
-     * type.
-     *
-     * <p>This method works even if the underlying Java type is a type variable. In that case, a
-     * 'null' AnnnotationMirror and the empty set represent a meaningful value (namely, no
-     * annotation).
+     * Returns true if and only if {@link AnnotatedTypeMirror#getAnnotations()} can return an empty
```

Comment:
Is it really whether it can be an empty set? Shouldn't it be whether fewer than the qualifier hierarchy width qualifiers appear on the type?
For example in nullness, a `<T> @Nullable T foo()...` is valid, even though the qualifier hierarchy has width 2.